app/models/concerns/omniauthable.rb

```diff
@@ -31,6 +31,27 @@ def find_for_oauth(auth, signed_in_resource = nil)
       user   = signed_in_resource || identity.user
       user ||= create_for_oauth(auth)
 
+      # get the roles claim from the extra info for this client id, if it is provided
+      # mastodon only supports a single role, so check in order of priority.
+      oidc_config = Devise.omniauth_configs[auth.provider.to_sym]
+      oidc_client_id = oidc_config.options[:client_options][:identifier]
+
+      roles = auth.dig 'extra', 'raw_info', 'resource_access', oidc_client_id, 'roles'
+      role_id = nil
+
+      if roles&.include? 'owner'
```

Comment:
I am unfamiliar with OIDC, are these role names hardcoded in OIDC, or are they variable?

This might require having a way to configure the mapping.

Reply:
the first two (`extra` and `raw_info`) are not OIDC specific, they have to do with poking into the omniauth structures (and would make more sense for that package to expose to the caller so that we don't have to dig around in the raw response)

the `resource_access` - `$oidc_client_id` - `roles` are part of the OIDC user information structure that Keycloak provides in the `userinfo` claim. I'm not certain if other OIDC providers use different names; that is typically mapped on the SSO provider side anyway.

Reply:
OIDC Core doesn't actually define RBAC mechanisms, there are `roles` and `groups` JWT claims extensions defined in [RFC 7643 - SCIM Core](https://www.rfc-editor.org/rfc/rfc7643.html#section-4.1.2), however this just defines them as arbitrary strings that would require mapping.

So you'd want to really have a `UserRole` with a "oidc_mapping", and then to find the highest permission role out of the potential roles.

The path being used here of `resource_access.${oidc_client_id}.roles` inside the `raw_info` is specific to how keycloak can serve RBAC to a client, and not standard to OIDC. For instance, Auth0 uses either the `roles` claim or the `scopes` field for presenting permissions, AWS Cognito uses the `cognito:roles` and `cognito:preferred_role` claims.

Perhaps the best option here would be to include a WebHook triggered with the ID Token that can the set the correct roles, if this is ultimately desired, that way a wide array of potential methods can be used.